app/models/miq_queue.rb

```diff
@@ -184,6 +184,237 @@ def self.submit_job(options)
     put(options)
   end
 
+  # options:
+  #   :service
+  #   :affinity    (optional)
+  #   :expires_on  (optional, timestamp in second)
+  #   :deliver_on  (optional, timestamp in second)
+  #   :priority    (optional, 0 - 9)
+  #   :class_name
+  #   :method_name
+  #   :instance_id (optional)
+  #   :args        (optional, an array)
+  #
+  # service is used to decide queue name.
+  def self.put_background_job(client = nil, options)
+    assert_options(options, [:class_name, :method_name, :service])
+
+    options = options.dup
+    address, headers = queue_for_publish(options)
+
+    unless client
+      client = ActiveMqClient.open
+      close_on_exit = true
+    end
+    client.publish(address, options.to_yaml, headers)
+    _log.info("Address(#{address}), msg(#{options.inspect}), headers(#{headers.inspect})")
+    client.close if close_on_exit
+  end
+
+  # options:
+  #   :service
+  #   :affinity (optional)
+  def self.subscribe_background_job(client, options)
+    assert_options(options, [:service])
+
+    options = options.dup
+    queue_name, headers = queue_for_subscribe(options)
+
+    client.subscribe(queue_name, headers) do |msg|
+      begin
+        msg_options = YAML.load(msg.body).with_indifferent_access
+
+        # TODO: refactor after MiqQueue.put and table miq_queues are deprecated
+        dequeue = MiqQueue.new(msg_options)
+
+        status, _message, result = dequeue.deliver(MiqServer.my_server || MiqServer.seed)
+      rescue => err
+        _log.error("Error delivering #{dequeue.inspect}, reason: #{err}")
+      end
+
+      if status == "timeout"
+        begin
+          _log.info("Reconnecting to DB after timeout error during queue deliver")
+          ActiveRecord::Base.connection.reconnect!
+        rescue => err
+          _log.error("Error encountered during <ActiveRecord::Base.connection.reconnect!> error:#{err.class.name}: #{err.message}")
+        end
+      end
+      # pp result
+
+      # No call back support yet
+      # queue.delivered(status, message, result) unless status == 'retry'
+
+      client.ack(msg)
+    end
+  end
+
+  # options:
+  #   :service
+  #   :affinity (optional)
+  #   :message
+  #   :sender (optional)
+  #   :message_type (optional)
+  #   :other keys (expiires_on, deliver_on, priority) same as put_background_job
+  #
+  # service is used to decide the queue name.
+  # message can be any object to be serialized through yaml.
+  def self.put_job(client = nil, options)
+    assert_options(options, [:message, :service])
+
+    options = options.dup
+    address, headers = queue_for_publish(options)
+    headers[:sender] = options.delete(:sender) if options[:sender]
+    headers[:message_type] = options.delete(:message_type) if options[:message_type]
+
+    unless client
+      client = ActiveMqClient.open
+      close_on_exit = true
+    end
+    client.publish(address, options[:message].to_yaml, headers)
```

Comment:
If we want the this smartness we need to pass on the data type so that the subscriber knows how to interpret the string. Always converting to yaml may be a waste sometimes but saves us some headache managing the types.